src/Npgsql/NpgsqlTypes/NpgsqlDate.cs

```diff
@@ -108,72 +108,204 @@ public NpgsqlDate(int year, int month, int day)
-                throw new ArgumentNullException(nameof(str));
-            }
-
-            if (str == "infinity")
+        /// <summary>
+        /// Parses the well-known text representation of a PostgreSQL date type into a <see cref="NpgsqlDate"/>.
+        /// </summary>
+        /// <param name="input">A PosgreSQL date type in a well-known text format.</param>
+        /// <returns>
+        /// The <see cref="NpgsqlDate"/> represented by the <paramref name="input"/>.
+        /// </returns>
+        /// <exception cref="FormatException">
+        /// Malformed date literal.
+        /// </exception>
+        /// <remarks>
+        /// See: https://www.postgresql.org/docs/current/static/datatype-datetime.html
+        /// </remarks>
+        public static NpgsqlDate Parse([NotNull] string input)
+        {
+            if (input is null)
+                throw new ArgumentNullException(nameof(input));
+
+            var str = input.AsSpan().Trim();
+
+            if (str.SequenceEqual(InfinityLiteral.AsSpan()))
                 return Infinity;
 
-            if (str == "-infinity")
+            if (str.SequenceEqual(NegativeInfinityLiteral.AsSpan()))
                 return NegativeInfinity;
 
-            str = str.Trim();
-            try {
-                var idx = str.IndexOf('-');
-                if (idx == -1) {
-                    throw new FormatException();
-                }
-                var year = int.Parse(str.Substring(0, idx));
-                var idxLast = idx + 1;
-                if ((idx = str.IndexOf('-', idxLast)) == -1) {
-                    throw new FormatException();
-                }
-                var month = int.Parse(str.Substring(idxLast, idx - idxLast));
-                idxLast = idx + 1;
-                if ((idx = str.IndexOf(' ', idxLast)) == -1) {
-                    idx = str.Length;
-                }
-                var day = int.Parse(str.Substring(idxLast, idx - idxLast));
-                if (str.Contains("BC")) {
+            try
+            {
+                var yearSeparator = str.IndexOf('-');
+
+                if (yearSeparator == -1)
+                    throw new FormatException($"Malformed date literal: {str.ToString()}");
+
+                var monthSeparator = str.Slice(yearSeparator + 1).IndexOf('-');
+
+                if (monthSeparator == -1)
+                    throw new FormatException($"Malformed date literal: {str.ToString()}");
+
+                var daySeparator = str.Slice(yearSeparator + monthSeparator + 1).IndexOf(' ');
+
+                var yearSegment = str.Slice(0, yearSeparator);
+                var monthSegment = str.Slice(yearSeparator + 1, monthSeparator);
+                var daySegment =
+                    daySeparator == -1
+                        ? str.Slice(yearSeparator + monthSeparator + 2)
+                        : str.Slice(yearSeparator + monthSeparator + 2, daySeparator);
+
+                var year = int.Parse(yearSegment.ToString());
```

Comment:
Doing `ToString()` here kind of defeats the purpose of using Span - we're allocating.

.NET Core 2.1 actually adds overloads that allow you to parse directly from a `ReadOnlySpan<char>` (see https://github.com/dotnet/corefx/issues/22403), but we don't have them in anything other than .NET Core 2.1. So if we want to do this right, we need to have Npgsql target netcoreapp21 and add an `#if` here, using the new overloads. This would provide the optimal zero-alloc perf for .NET Core 2.1, while still working for older targets. 

Otherwise I'm simply not sure that this PR is really worth it...

Reply:
Ah. I figured I had something misconfigured on my end that was hiding those overloads, and went with `ToString()` as a bandaid at the time. I'll add the `#if` blocks in the next push.